### ISSUE TITLE
Negative spread on the DEX with offers not matching

### ISSUE THREAD

**Issue by KoenPaas** [ripple/rippled#4081]
## Issue Description
Users are reporting negative spreads on the XRPL DEX, this could be either an issue with the `book_offers` method or offers not matching.

## Steps to Reproduce
Have not been able to reproduce.

## Expected Result
That these orders match and fill

## Actual Result
When getting the `book_offers` there are offers that do not match in the orderbook and stay active effectively freezing trading for that currency pair. 

## Supporting Files

![2c834f11-44e4-4d5e-b8a5-16b812923678](https://user-images.githubusercontent.com/36813373/150768396-b88e1ef1-bc81-4852-8476-ddf2bfac28c2.png)

![eea9bce7-a1c7-4fd7-9e28-2df3675980ae](https://user-images.githubusercontent.com/36813373/150768420-02f5638f-4276-4863-a84e-401252af1624.jpeg)

**Comment by seelabs** [comment]
Thanks for the report @KoenPaas. I'll look into this. Can you let me know who the issuer is for the SGB side of the book? I can't tell from the screenshots.

**Comment by KoenPaas** [comment]
Thank you,  The issuer is GateHub, rctArjqVvTHihekzDeecKo6mkTYTUSBNc the reports are from last friday and saterday

**Comment by wojake** [comment]
This happened to me as well: [here](https://www.xrpchat.com/topic/38893-offercreate-buy-needs-sell-but-not-the-other-way-around/?do=findComment&comment=961937)

**Comment by seelabs** [comment]
I been working on this issue and I wanted to give some status.

Since this issue reported a strange ledger state, rather than a strange transaction, the first task was to try to find transactions that left the order book with a negative spread. I wrote a script to search the ledger from Thursday, Jan 20 to Sunday, Jan 23 I found one transaction:

```
AD5A0626BC4EBE8FD032AC728667CFC2DD00469A096B39B3F51CCA3E70733497
```

After this transaction runs, the XRP and SGB currency pair had a negative spread. The script did not find any other transactions that caused a negative spread.

Replaying this transaction, we see the initial quality is .155 (`taker pays 5000 SGP/taker gets 775 XRP`). It consumes two offers in the order book and tries to consume an offer with a rate of .155 (`taker pays 308.14 XRP` and `taker gets 1988 SGP`. This has a rate or .155 (`taker pays 308.14 XRP/taker gets 1988SGP`). Since the offer was not placed passively, this offer should have been crossed. The reason it wasn't is the payment engine only consumed part of it, and due to rounding the rate for partial offer was computed to be slightly larger than our .155 cutoff, so it did not cross.

At the end of the transaction, a new offer is created with a quality of `.155000000280508`, a tiny bit above the original .155. Since quantities aren't infinitely divisible, there is going to be small differences in quality when we consume offers. This is normal. The tip of the SGP to XRP order book now has a quality slightly about .155 and the tip of the XRP to SGP has a rate (1/quality) of exactly .155, so we get a (very small) negative spread.

That describes what happened in that transaction. However, I'm not convinced that this is the issue being reported here. The screen shot is showing a large negative spread, while this transaction only explains a small negative spread.

I know this doesn't resolve the issue, but I did want to describe what I've found so far. What I really need to find is some transaction that leaves a large negative spread and replay that, and the script I wrote didn't find one. It's possible there's a bug in the script (I'll ask someone on the team to review to). I'll also try to think of other ways to try to get to the bottom on this.

**Comment by KoenPaas** [comment]
Thank you for putting in this effort @seelabs

**Comment by Cadey** [comment]
I placed a buit report this here

https://github.com/ripple/rippled/issues/3985

When this bug is on the ledgeer an even worse situation is possible. When you end up with the negative spread you can place a "BUY" order for any amount and if there is a sell order it will match it. Essentially allowing a sell order high up on the border book to full fill before it should

**Comment by lathanbritz** [comment]
This issue is most likely a result of stale offer objects in the ledger. book_offer returns all orders and does not filter the expired timestamp on the data returned. This needs to be completed by the interface consuming the data. Tucked away under the explanation of the Expire field in the documentation this is explained https://xrpl.org/offers.html#offer-expiration

Here is a git that shows how to filter. https://gist.github.com/lathanbritz/76c47806218691f3401486aa6a6e5eb0

**Comment by Administratordes** [comment]
Hello GoodDay if you need any help or assistance regarding to any issue at all please don’t hesitate to contact the LiveChat Support https://direct.lc.chat/14065185/

**Comment by yyolk** [comment]
X-referencing the standard amendment proposal here https://github.com/XRPLF/XRPL-Standards/discussions/74

**Comment by yyolk** [comment]
X-referencing the PR here: https://github.com/ripple/rippled/pull/4164

**Comment by julianwells3510** [comment]
Yes, there’s some kind of givеаway going on. I got some Ethereum there yesterday for different wallets. It’s unlikely they’ll take it back from me, but it’s a pity I didn’t have more wallets and couldn’t get more. But maybe it’s for the best.

<img width="369" alt="Image" src="https://github.com/user-attachments/assets/1aa26a32-6f8f-4eeb-9598-ec0263f89968" />

<img width="373" alt="Image" src="https://github.com/user-attachments/assets/9012d8a8-6501-4c63-80a6-f5e142af57db" />
